src/Microsoft.CodeAnalysis.ResxSourceGenerator/Microsoft.CodeAnalysis.ResxSourceGenerator/AbstractResxGenerator.cs

```diff
@@ -22,166 +23,164 @@
-            {
-                try
+            var resourceFiles = context.AdditionalTextsProvider.Where(static file => file.Path.EndsWith(".resx", StringComparison.OrdinalIgnoreCase));
+            var compilationInformation = context.CompilationProvider.Select(
+                (compilation, cancellationToken) =>
                 {
-                    ProcessResourceFile(context, resourceFile);
-                }
-                catch (Exception ex)
+                    var methodImplOptions = compilation.GetOrCreateTypeByMetadataName(typeof(MethodImplOptions).FullName);
+                    var hasAggressiveInlining = methodImplOptions?.MemberNames.Contains(nameof(MethodImplOptions.AggressiveInlining)) ?? false;
+                    var hasNotNullIfNotNull = compilation.GetOrCreateTypeByMetadataName("System.Diagnostics.CodeAnalysis.NotNullIfNotNullAttribute") is not null;
+
+                    return new CompilationInformation(
+                        AssemblyName: compilation.AssemblyName,
+                        CodeLanguage: compilation.Language,
+                        SupportsNullable: SupportsNullable(compilation),
+                        HasAggressiveInlining: hasAggressiveInlining,
+                        HasNotNullIfNotNull: hasNotNullIfNotNull);
+                });
+            var resourceFilesToGenerateSource = resourceFiles.Combine(context.AnalyzerConfigOptionsProvider).Combine(compilationInformation).SelectMany(
+                static (resourceFileAndOptions, cancellationToken) =>
                 {
-                    var exceptionLines = ex.ToString().Split(new[] { Environment.NewLine }, StringSplitOptions.None);
-                    var text = string.Join("", exceptionLines.Select(line => "#error " + line + Environment.NewLine));
-                    var errorText = SourceText.From(text, Encoding.UTF8, SourceHashAlgorithm.Sha256);
-                    context.AddSource($"{Path.GetFileName(resourceFile.Path)}.Error", errorText);
-                }
-            }
-        }
-
-        private void ProcessResourceFile(GeneratorExecutionContext context, AdditionalText resourceFile)
-        {
-            var options = context.AnalyzerConfigOptions.GetOptions(resourceFile);
+                    var ((resourceFile, optionsProvider), compilationInfo) = resourceFileAndOptions;
+                    var options = optionsProvider.GetOptions(resourceFile);
+                    if (options.TryGetValue("build_metadata.AdditionalFiles.GenerateSource", out var generateSourceText)
+                        && !string.IsNullOrEmpty(generateSourceText)
+                        && generateSourceText != "unset"
+                        && !string.Equals(bool.TrueString, generateSourceText, StringComparison.OrdinalIgnoreCase))
+                    {
+                        // Source generation is disabled for this resource file
+                        return Array.Empty<ResourceInformation>();
+                    }
```

Comment:
this part seems to not depend on compilationInfo at all.  *perhaps* have a `where` on the combined resourceFile+optionsProvider provider that filters this case out, and only continues on if this isn't present.

Reply:
`SelectMany` is being used as a combination of `Where` and `Select` so the input only needs to be processed once. Note that this method operates on a single resource file, so the processing of that resource file here will _either_ exclude it _or_ gather information about it. Changes to one resource file will not require re-processing the other resource files through this step.